app/src/main/java/com/github/onlynotesswent/utils/NotesToFlashcard.kt

```diff
@@ -74,25 +111,219 @@ class NotesToFlashcard(
+   * @param onFailure Callback invoked when the conversion process fails with an exception.
+   */
+  fun convertFolderToDecks(onSuccess: (Deck) -> Unit, onFailure: (Exception) -> Unit) {
+    val currentFolder = folderViewModel.selectedFolder.value
+    if (currentFolder == null) {
+      onFailure(IllegalStateException("No folder selected"))
+      return
+    }
+
+    getOrCreateDeckSubFolder(
+        currentFolder,
+        null,
+        onSuccess = { deckFolder ->
+          processFolderRecursively(currentFolder, deckFolder, onSuccess, onFailure)
+          // Reset sub folder list and folder notes
+          folderViewModel.getSubFoldersOf(parentFolderId = currentFolder.id)
+          noteViewModel.getNotesFromFolder(folderId = currentFolder.id)
+        },
+        onFailure = { error ->
+          Log.e(
+              TAG, "Failed to create or retrieve root deck folder for ${currentFolder.name}", error)
+          onFailure(error)
+        })
+  }
+
+  /**
+   * Processes a folder and its subfolders recursively to convert notes into decks of flashcards.
+   *
+   * This method retrieves notes and subfolders within a given folder. Each note is converted into a
+   * deck of flashcards, and each subfolder is recursively processed. Combined decks are created if
+   * a folder contains multiple notes or subfolders.
+   *
+   * @param folder The folder to process.
+   * @param deckFolder The deck folder where decks will be stored.
+   * @param onSuccess Callback invoked when the processing is successful with the final deck.
+   * @param onFailure Callback invoked when the processing fails with an exception.
+   */
+  private fun processFolderRecursively(
+      folder: Folder,
+      deckFolder: Folder,
+      onSuccess: (Deck) -> Unit,
+      onFailure: (Exception) -> Unit
+  ) {
+    noteViewModel.getNotesFromFolder(
+        folder.id,
+        onSuccess = { notes ->
+          val folderFlashcardIds = mutableListOf<String>()
+
+          // Convert notes to flashcards
+          notes.forEach { note ->
+            convertNoteToDeck(
+                note,
+                folderId = deckFolder.id,
+                onSuccess = { deck -> folderFlashcardIds.addAll(deck.flashcardIds) },
+                onFileNotFoundException = { Log.e(TAG, "File not found for note ${note.id}") },
+                onFailure = { Log.e(TAG, "Failed to convert note ${note.id} to deck", it) })
+          }
+
+          // Process subfolders recursively
+          folderViewModel.getSubFoldersOf(
+              parentFolderId = folder.id,
+              onSuccess = { subfolders ->
+                subfolders.forEach { subfolder ->
+                  getOrCreateDeckSubFolder(
+                      subfolder,
+                      deckFolder,
+                      onSuccess = { deckSubFolder ->
+                        processFolderRecursively(subfolder, deckSubFolder, onSuccess, onFailure)
```

Comment:
I didn't think about the use case of the onSuccess yet. Was planning on modifying it later when doing the ui. However, now I modified so that onSuccess is called correctly with only the last deck.